zeebe/clients/java/src/test/java/io/camunda/zeebe/client/util/RestGatewayService.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.zeebe.client.util;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.tomakehurst.wiremock.client.WireMock;
+import com.github.tomakehurst.wiremock.junit5.WireMockRuntimeInfo;
+import io.camunda.zeebe.client.protocol.rest.BrokerInfo;
+import io.camunda.zeebe.client.protocol.rest.Partition;
+import io.camunda.zeebe.client.protocol.rest.Partition.HealthEnum;
+import io.camunda.zeebe.client.protocol.rest.Partition.RoleEnum;
+import io.camunda.zeebe.client.protocol.rest.TopologyResponse;
+import java.util.Arrays;
+import org.junit.jupiter.api.Assertions;
+
+public class RestGatewayService {
+
+  public static final ObjectMapper JSON_MAPPER = new ObjectMapper();
+
+  private final WireMockRuntimeInfo mockInfo;
+
+  public RestGatewayService(final WireMockRuntimeInfo mockInfo) {
+    this.mockInfo = mockInfo;
+  }
+
+  public void onTopologyRequest(
+      final int clusterSize,
+      final int partitionsCount,
+      final int replicationFactor,
+      final String gatewayVersion,
+      final BrokerInfo... brokers) {
+    try {
+      mockInfo
+          .getWireMock()
+          .register(
+              WireMock.get("/v2/topology")
+                  .willReturn(
+                      WireMock.okJson(
+                          JSON_MAPPER.writeValueAsString(
+                              new TopologyResponse()
+                                  .gatewayVersion(gatewayVersion)
+                                  .clusterSize(clusterSize)
+                                  .replicationFactor(replicationFactor)
+                                  .partitionsCount(partitionsCount)
+                                  .brokers(Arrays.asList(brokers))))));
```

Comment:
🔧 I want to suggest moving the construction of the response object to a parameter of the `onTopologyRequest` method. This would result in improved readability in the test cases:

```java
gatewayService.onTopologyRequest(
  new TopologyResponse()
     .gatewayVersion("1.22.3-SNAPSHOT")
     .clusterSize(2)
     .replicationFactor(3)
     .partitionsCount(10)
     .brokers(Arrays.asList(
        broker(
            0,
            "host1",
            123,
            "1.22.3-SNAPSHOT",
            partition(0, LEADER, HEALTHY),
            partition(1, FOLLOWER, UNHEALTHY)),
        broker(
            1,
            "host2",
            212,
            "2.22.3-SNAPSHOT",
            partition(0, FOLLOWER, HEALTHY),
            partition(1, LEADER, HEALTHY)),
        broker(
            2,
            "host3",
            432,
            "3.22.3-SNAPSHOT",
            partition(0, FOLLOWER, UNHEALTHY),
            partition(1, INACTIVE, UNHEALTHY)))));
```

I'm suggesting this, because this test case and this `onX()` method will probably become an example for other tests.

🔧 To further accentuate this becoming a new API, you could extract an interface that is made available from the `ClientRestTest`, that also adds some JavaDoc to these `onX` methods.

Reply:
Thanks, @korthout 👍 

> I want to suggest moving the construction of the response object to a parameter of the onTopologyRequest method.

That would also help limit the responsibility of this class to only handling requests, not serving as a helper to create them as well. I'll go for that, should make it clearer and more readable.

> To further accentuate this becoming a new API, you could extract an interface that is made available from the ClientRestTest, that also adds some JavaDoc to these onX methods.

Adding Javadoc to the class itself will be most helpful here, I think. There will be no other implementation of this interface, as far as I can see. I am rather wondering if the `ClientRestTest` itself could offer the `onX()` methods to make them more discoverable while writing tests. That would remove the `RestGatewayService` altogether.

What do you think?

Reply:
>I am rather wondering if the ClientRestTest itself could offer the onX() methods to make them more discoverable while writing tests.

I can see the upside, but it comes at the cost of the `ClientRestTest` doing multiple things:
- providing the test setup in the first place
- allowing to mock endpoints

My feeling is that splitting the responsibility is easier to maintain as the number of endpoints grows. But it's more a gut feeling, than something I can argue for easily. Feel free to find an approach that you think fits well.

Reply:
Good points, yes.

My proposal would be:
* Let's separate request handling and request creation (tests create requests, the service handles them).
* Let's keep the separation between test setup (in `ClientRestTest`) and request handling (in `RestGatewayService`)
* Let's add Javadoc to the service's `onX()` methods to set a good example.